File: problem_7.cpp
#include <chrono>
#include <iostream>
#include <vector>
#include <cstring>
#include <cmath>

const long limitNumber = 120000;
const int soughtPrime = 10001;

std::vector<long> primes;

std::vector<long>* allPrimesUsingSieve (long n) 
{
    std::vector<unsigned char> sieve;
    if (primes.size() > 0) 
    {
        primes.resize(0);
    }
    sieve.resize(n + 1);
    memset(sieve.data(), 0, (n + 1) * sizeof(unsigned char));
    for (long i = 2; i < int(sqrt(n)); i++) 
    {
        if (sieve[i] != 0) 
        {
            continue;
        }
        for (long j = i + i; j < n; j += i) 
        {
            sieve[j] = (unsigned short) 1;
        }
    }
    for (long i = 2; i <= n; i++) 
    {
        if (sieve[i] == 0)
        {
            primes.push_back(i);
        }
    }
    return &primes;
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    long prime = 0;
    std::vector<long>* primes = allPrimesUsingSieve(limitNumber);
    if (primes->size() >= soughtPrime) 
    {
        prime = primes->at(soughtPrime - 1);
    }
    
    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << prime << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_7.rs
use std::time::{Instant};

const LIMIT_NUMBER: usize = 120000;
const SOUGHT_PRIME: usize = 10001;

fn all_primes_using_sieve (n: usize) -> Vec<usize>
{
    let mut sieve: Vec<u8> = vec![];
    let mut primes: Vec<usize> = vec![];
    sieve.resize(n + 1, 0);
    let limit = f64::sqrt(n as f64) as usize;
    for i in 2..limit 
    {
        if sieve[i] == 0 
        {
            for j in ((i + i)..n).step_by(i)
            {
                sieve[j] = 1;
            }
        }
    }
    for i in 2..n 
    {
        if sieve[i] == 0
        {
            primes.push(i);
        }
    }
    return primes;
}

fn main () 
{
    let start = Instant::now();

    let all_primes = all_primes_using_sieve(LIMIT_NUMBER);
    if all_primes.len() >= SOUGHT_PRIME
    {
        println!("Result: {}", all_primes[SOUGHT_PRIME  - 1]);
    }

    let duration = start.elapsed();
    println!("Timing: {}", duration.as_nanos());
}

